File: Go/Tester.go
package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"reflect"
	"time"
)

func main() {
	f, err := os.Open("../sudoku.csv")
	if err != nil {
		fmt.Println(err)
		return
	}
	defer f.Close()

	csvReader := csv.NewReader(f)
	line, err := csvReader.Read()
	line, err = csvReader.Read()
	index:= 1
	block := 1
	overall := time.Now()
	start := time.Now()
	for err == nil {
		quizUnformatted:= line[0]
		var quiz [9][9]int
		var solution [9][9]int
		solutionUnformatted:= line[1]
		for i:= 0; i < 9; i++ {
			for j:= 0; j < 9; j++ {
				quiz[i][j]= int(quizUnformatted[i*9+j]) - '0'
				solution[i][j]= int(solutionUnformatted[i*9+j]) - '0'
			}
		}
		result := Solve(quiz)
		if !reflect.DeepEqual(result, solution) {
			fmt.Printf("Error on quiz %d\n", index)
			fmt.Printf("Answer found: %v\n", result)
			fmt.Printf("Expected answer: %v\n", solution)
			break
		}
		if index % 50000 == 0 && index >= 50000{
			duration := time.Now().Sub(start)
			overallDuration := time.Now().Sub(overall)
			minutes := int(duration.Minutes())
			seconds := int(duration.Seconds()) % 60
			milliseconds := int(duration.Milliseconds()) % 1000
			overallMinutes := int(overallDuration.Minutes())
			overallSeconds := int(overallDuration.Seconds()) % 60
			overallMilliseconds := int(overallDuration.Milliseconds()) % 1000
			fmt.Printf("Block %d took %d minutes, %d seconds, and %d milliseconds.\n", 
			block, minutes, seconds, milliseconds)
			start = time.Now()
			block++
			if index != 1000000 {
				fmt.Printf("Solved %d quizzes in %d minutes, %d seconds, and %d milliseconds.\n", 
				index, overallMinutes, overallSeconds, overallMilliseconds)
			} else {
				fmt.Printf("Overall time: %d minutes, %d seconds, and %d milliseconds.\n",
				overallMinutes, overallSeconds, overallMilliseconds)
			}
		}
		line, err = csvReader.Read()
		index++
	}
}

File: C++/Tester.cpp
//
// Created by Patrick on 7/9/2022.
//
#include "Solver.h"
#include <fstream>
#include <iostream>
#include <sstream>
#include <time.h>

using namespace std;


int main(){
    ifstream infile("../sudoku.csv");
    string line;
    getline(infile, line);
    int index = 1;
    int block = 1;
    auto start = clock();
    auto overall_start = clock();
    while (getline(infile, line)) {
        int quiz[81];
        int answer[81];
        auto pos = line.find(",");
        string quizLine = line.substr(0, pos);
        string answerLine = line.substr(pos + 1);
        for (int i = 0; i < 81; i++) {
            quiz[i] = quizLine[i] - '0';
            answer[i] = answerLine[i] - '0';
        }
        solve(quiz);
        for (int i = 0; i < 81; i++) {
            if (quiz[i] != answer[i]) {
                cout << "Error on Quiz " << index << endl;
                ostringstream os;
                for (int c: quiz) {
                    os << c;
                }
                cout << "Solution found: " << os.str() << endl;
                ostringstream os2;
                for (int c: answer) {
                    os2 << c;
                }
                cout << "Solution expected: " << os2.str() << endl;
                return 1;
            }
        }
        if (index % 50000 == 0 && index>=50000) {
            auto end = clock();
            auto milli = (end - start) % 1000;
            auto seconds = (end - start) / 1000;
            cout << "Block " << block << " took " << seconds << " seconds and " << milli << " milliseconds" << endl;
            milli = (clock() - overall_start) % 1000;
            seconds = (clock() - overall_start) / 1000;
            cout << "Solved " << index << " quizzes in " << seconds << " seconds and " << milli << " milliseconds" << endl;
            block++;
            start = clock();
        }
        index++;
    }
    cout << "All quizzes passed" << endl;
    infile.close();
    return 0;
}